arangod/Cluster/ClusterInfo.cpp

```diff
@@ -513,6 +515,124 @@ void ClusterInfo::triggerBackgroundGetIds() {
+  for (auto const& spec : replicatedStates) {
+    futureStates.emplace_back(
+        replicatedStateMethods->waitForStateReady(spec.id, *spec.version));
+  }
+
+  return futures::collectAll(std::move(futureStates))
+      .thenValue(
+          [&clusterInfo = _server.getFeature<ClusterFeature>().clusterInfo()](
+              auto&& raftIndices) {
+            consensus::index_t highestRaftIndex = std::transform_reduce(
+                raftIndices.begin(), raftIndices.end(), (consensus::index_t)0,
+                [](consensus::index_t l, consensus::index_t r) {
+                  return std::max(l, r);
+                },
+                [](auto&& value) { return value.get().get(); });
+
+            return clusterInfo.waitForPlan(highestRaftIndex);
+          })
+      .then([](auto&& tryResult) {
+        auto result =
+            basics::catchToResult([&] { return std::move(tryResult.get()); });
+        if (result.fail()) {
+          if (result.is(TRI_ERROR_NO_ERROR)) {
+            result = Result(TRI_ERROR_INTERNAL);
+          }
+
+          result = result.mapError([](result::Error error) {
+            error.appendErrorMessage(
+                "Failed to create a corresponding replicated state "
+                "for each shard!");
+            return error;
+          });
+        }
+
+        return result;
+      });
+}
+
+auto ClusterInfo::deleteReplicatedStates(
+    std::string const& databaseName,
+    std::vector<replication2::LogId> const& replicatedStatesIds)
+    -> futures::Future<Result> {
+  auto replicatedStateMethods =
+      arangodb::replication2::ReplicatedStateMethods::createInstanceCoordinator(
+          _server, databaseName);
+
+  std::vector<futures::Future<Result>> deletedStates;
+  deletedStates.reserve(replicatedStatesIds.size());
+  for (auto const& id : replicatedStatesIds) {
+    deletedStates.emplace_back(
+        replicatedStateMethods->deleteReplicatedState(id));
+  }
+
+  return futures::collectAll(std::move(deletedStates))
+      .then([](futures::Try<std::vector<futures::Try<Result>>>&& tryResult) {
+        auto deletionResults =
+            basics::catchToResultT([&] { return std::move(tryResult.get()); });
+
+        auto result = deletionResults.result();
+        if (result.fail()) {
+          return result.mapError([](result::Error error) {
+            error.appendErrorMessage(
+                "Failed to delete replicated states corresponding to shards!");
+            return error;
+          });
+        }
+        for (auto shardResult : deletionResults.get()) {
+          auto r = basics::catchToResult(
+              [&] { return std::move(shardResult.get()); });
+          if (r.fail()) {
+            return result.mapError([](result::Error error) {
+              error.appendErrorMessage(
+                  "Failed to delete replicated states corresponding to "
+                  "shards!");
+              return error;
+            });
+          }
+        }
```

Comment:
We can probably reduce some code duplication here.

Reply:
Write the `r.fail()` part in a lambda and use the lambda twice?

Reply:
Yes, exactly...